ssl/rio/poll_builder.c

```diff
@@ -0,0 +1,170 @@
+#include "poll_builder.h"
+#include "internal/safe_math.h"
+#include <assert.h>
+
+OSSL_SAFE_MATH_UNSIGNED(size_t, size_t)
+
+int ossl_rio_poll_builder_init(RIO_POLL_BUILDER *rpb)
+{
+#if RIO_POLL_METHOD == RIO_POLL_METHOD_SELECT
+    FD_ZERO(&rpb->rfd);
+    FD_ZERO(&rpb->wfd);
+    FD_ZERO(&rpb->efd);
+    rpb->hwm_fd     = -1;
+#elif RIO_POLL_METHOD == RIO_POLL_METHOD_POLL
+    rpb->pfd_heap   = NULL;
+    rpb->pfd_num    = 0;
+    rpb->pfd_alloc  = OSSL_NELEM(rpb->pfds);
+#endif
+    return 1;
+}
+
+void ossl_rio_poll_builder_cleanup(RIO_POLL_BUILDER *rpb)
+{
+    if (rpb == NULL)
+        return;
+
+#if RIO_POLL_METHOD == RIO_POLL_METHOD_POLL
+    OPENSSL_free(rpb->pfd_heap);
+#endif
+}
+
+#if RIO_POLL_METHOD == RIO_POLL_METHOD_POLL
+static int rpb_ensure_alloc(RIO_POLL_BUILDER *rpb, size_t alloc)
+{
+    struct pollfd *pfd_heap_new;
+    size_t total_size;
+    int error = 0;
+
+    if (alloc <= rpb->pfd_alloc)
+        return 1;
+
+    total_size = safe_mul_size_t(alloc, sizeof(struct pollfd), &error);
+    if (error)
+        return 0;
+
+    pfd_heap_new = OPENSSL_realloc(rpb->pfd_heap, total_size);
+    if (pfd_heap_new == NULL)
+        return 0;
+
+    rpb->pfd_heap   = pfd_heap_new;
+    rpb->pfd_alloc  = alloc;
+    return 1;
+}
+#endif
+
+int ossl_rio_poll_builder_add_fd(RIO_POLL_BUILDER *rpb, int fd,
+                                 int want_read, int want_write)
+{
+    if (fd < 0)
+        return 0;
+
+#if RIO_POLL_METHOD == RIO_POLL_METHOD_SELECT
+
+# ifndef OPENSSL_SYS_WINDOWS
+    /*
+     * On Windows there is no relevant limit to the magnitude of a fd value (see
+     * above). On *NIX the fd_set uses a bitmap and we must check the limit.
+     */
+    if (fd >= FD_SETSIZE)
+        return 0;
+# endif
+
+    if (want_read)
+        openssl_fdset(fd, &rpb->rfd);
+
+    if (want_write)
+        openssl_fdset(fd, &rpb->wfd);
+
+    openssl_fdset(fd, &rpb->efd);
+    if (fd > rpb->hwm_fd)
+        rpb->hwm_fd = fd;
+
+    return 1;
+
+#elif RIO_POLL_METHOD == RIO_POLL_METHOD_POLL
+    size_t num_loop;
+
+    for (num_loop = 0;; ++num_loop) {
+        size_t i;
+        struct pollfd *pfds = (rpb->pfd_heap != NULL ? rpb->pfd_heap : rpb->pfds);
+        struct pollfd *pfd;
+
+        for (i = 0; i < rpb->pfd_num; ++i) {
+            pfd = &pfds[i];
+
+            if (pfd->fd == -1 || pfd->fd == fd)
+                break;
+        }
+
+        if (i >= rpb->pfd_alloc) {
+            /* Was not able to find room - enlarge and try again. */
+            if (num_loop > 0)
+                /* Should not happen twice. */
+                return 0;
+
+            if (!rpb_ensure_alloc(rpb, rpb->pfd_alloc * 2))
+                return 0;
+
+            continue;
```

Comment:
It seems strange that we have to redo the `for (i = 0...` loop above in this case. We will not get a different answer for the value of `i`

Reply:
Yeah, that seems unnecessary.